src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Reliability/SqlConfigurableRetryLogicManager.CreateProvider.cs

```diff
@@ -0,0 +1,334 @@
+            // the only acceptable constructor is the default constructor.
+            object obj = Activator.CreateInstance(type);
+            SqlClientEventSource.Log.TryTraceEvent("<sc.{0}.{1}|INFO> An instance of `{2}` type is created to invoke the `{3}` method.",
+                                                   s_typeName, methodName, type.FullName, method.Name);
+            return method.Invoke(obj, funcParams);
+        }
+
+        private static object[] PrepareParamValues(ParameterInfo[] parameterInfos, SqlRetryLogicOption option, string retryMethod)
+        {
+            // The retry method must have at least one `SqlRetryLogicOption`
+            if (parameterInfos.FirstOrDefault(x => x.ParameterType == typeof(SqlRetryLogicOption)) == null)
+            {
+                string message = $"Failed to create {nameof(SqlRetryLogicBaseProvider)} object because of invalid {retryMethod}'s parameters." +
+                    $"{Environment.NewLine}The function must have a paramter of type '{nameof(SqlRetryLogicOption)}'.";
+                throw new InvalidOperationException(message);
+            }
+
+            object[] funcParams = new object[parameterInfos.Length];
+            for (int i = 0; i < parameterInfos.Length; i++)
+            {
+                var paramInfo = parameterInfos[i];
+
+                // Create parameters with default values that are not as SqlRetryLogicOption type.
+                if (paramInfo.HasDefaultValue && paramInfo.ParameterType != typeof(SqlRetryLogicOption))
+                {
+                    funcParams[i] = paramInfo.DefaultValue;
+                }
+
+                // Assign the 'option' object to the first parameter with 'SqlRetryLogicOption' type
+                // neither it doesn't have default value 
+                // or there isn't another parameter with the same type and without a default value.
+                else if (paramInfo.ParameterType == typeof(SqlRetryLogicOption))
+                {
+                    if (!paramInfo.HasDefaultValue
+                        || (paramInfo.HasDefaultValue
+                            && parameterInfos.FirstOrDefault(x => x != paramInfo && !x.HasDefaultValue && x.ParameterType == typeof(SqlRetryLogicOption)) == null))
+                    {
+                        funcParams[i] = option;
+                    }
+                    else
+                    {
+                        funcParams[i] = paramInfo.DefaultValue;
+                    }
+                }
+                else
+                {
+                    string message = $"Failed to create {nameof(SqlRetryLogicBaseProvider)} object because of invalid {nameof(retryMethod)}'s parameters."
+                        + $"{Environment.NewLine}Parameter '{paramInfo.ParameterType.Name} {paramInfo.Name}' doesn't have default value.";
+                    throw new InvalidOperationException(message);
```

Comment:
All exception messages need to come from Resource Builder (Strings.resx) and preferably built via SqlUtil.